### ISSUE TITLE
A clear definition for out-of-range dense field accesses

### ISSUE THREAD

**Issue by BillXu2000** [taichi-dev/taichi#2766]
Out-of-range sparse field accesses are well defined and will return a constant value, while behavior for out-of-range dense fields accesses is not clearly defined yet.
It seems that such accesses are now handled in a "circular array" style: `a[x]` is in fact `a[x mod n]`, where n is the size of array `a`.
The difference between out-of-range accesses in sparse and dense field might be confusing for users, and my opinion is that we can make a clear definition in Taichi Document.

**Concisely describe the proposed feature**
A clear and concise description of what you want. For example,
> I would like to add an Apple Metal backend to the compiler so that my Macbook GPU can be utilized.

**Describe the solution you'd like (if any)**
A clear and concise description of what you want to achieve and implement.

**Additional comments**

**Comment by KLozes** [comment]
> It seems that such accesses are now handled in a "circular array" style: a[x] is in fact a[x mod n], where n is the size of array a.

Cool observation! This means Taichi automatically satisfies periodic boundary conditions in stencil operations! It is definitely a good idea to document this. 

> Out-of-range sparse field accesses are well defined and will return a constant value, while the behavior for out-of-range dense fields accesses is not clearly defined yet

Its seems to me that sparse fields are behaving the same way as dense? Here's a test.

```python
import taichi as ti

ti.init(arch=ti.cuda)

x = ti.field(ti.f32)
y = ti.field(ti.f32)
n = 4

ti.root.pointer(ti.ij, n).dense(ti.ij, 4).place(x, y)

@ti.kernel
def init_x():
    for i,j in ti.ndrange(n*4,n*4):
        x[i,j] = i

@ti.kernel
def init_y():
    for i,j in y:
        y[i,j] = x[i,j] + x[i,j+1] + x[i,j-1]

init_x()
init_y()

for j in range(4*n):
    for i in range(4*n):
        print(y[i,j])
        assert(y[i,j] == 3 * i)
```

**Comment by k-ye** [comment]
> Its seems to me that sparse fields are behaving the same way as dense?

You're right! I probably misguided @BillXu2000  here a bit. The ambient value is used for sparse cells that are **not activated**, not out-of-boundary.

It could be useful to add more modes to process the boundaries, much like an OpenGL texture sampler. WDYT? @KLozes

**Comment by BillXu2000** [comment]
> Its seems to me that sparse fields are behaving the same way as dense? Here's a test.
I think you are right, thanks for the test.
I tried to say " Sparse field accesses on inactive elements are well defined and will return a constant value".
Corner cases for sparse field is easy to handle by extending the range of pointer SNode.

**Comment by KLozes** [comment]
> It could be useful to add more modes to process the boundaries, sort of like an OpenGL texture sampler. WDYT? @KLozes

That would be great! I see the periodic boundary we see here is just the GL_REPEAT option https://learnopengl.com/Getting-started/Textures.

**Comment by KLozes** [comment]
Here's a better test. Also, it fails `packed=True`!

``` python 
import taichi as ti

ti.init(arch=ti.gpu)

x = ti.field(ti.f32)
y = ti.field(ti.f32)
z = ti.field(ti.f32)
n = 8

ti.root.dense(ti.ij, n).place(x, y, z)


@ti.kernel
def init():
    for i, j in ti.ndrange(n, n):
        x[i, j] = i*j

@ti.kernel
def func():
    for i, j in x:
        y[i, j] = x[i+1, j] - x[i-1, j]
        z[i, j] = x[i, j+1] - x[i, j-1]

init()
func()

for j in range(n):
    for i in range(n):
        if i == 0 or i == n-1:
            assert(y[i, j] == - j * (n-2))
        else:
            assert(y[i, j] == j * 2)

        if j == 0 or j == n-1:
            assert(z[i, j] == - i * (n-2))
        else:
            assert(z[i, j] == i * 2)
```

**Comment by strongoier** [comment]
Out-of-range accesses are invalid. You might be able to run your sample programs under non-packed mode with power-of-two shape due to overflow. However, if you open `debug=True`, you will see error messages regarding out of range.
